### PR TITLE
JENA-1488: add a selective folding analyzer

### PR DESCRIPTION
**Pull request by kinow**
This PR adds a selective folding analyzer, as explained in JENA-1488.

It takes a list of characters, used as a white list. Everything that is not in the white list, gets oassed though the existing ASCIIFoldingFilter.

### CODE REVIEW

**Comment by kinow** [comment]
Example configuration used for testing:

```
@prefix :        <#> .
@prefix fuseki:  <http://jena.apache.org/fuseki#> .
@prefix dc:      <http://purl.org/dc/elements/1.1/> .
@prefix rdf:     <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix rdfs:    <http://www.w3.org/2000/01/rdf-schema#> .
@prefix tdb:     <http://jena.hpl.hp.com/2008/tdb#> .
@prefix ja:      <http://jena.hpl.hp.com/2005/11/Assembler#> .
@prefix text:    <http://jena.apache.org/text#> .
@prefix skos:    <http://www.w3.org/2004/02/skos/core#> .

[] ja:loadClass "org.apache.jena.tdb.TDB" .
tdb:DatasetTDB  rdfs:subClassOf  ja:RDFDataset .
tdb:GraphTDB    rdfs:subClassOf  ja:Model .

[] ja:loadClass "org.apache.jena.query.text.TextQuery" .
text:TextDataset      rdfs:subClassOf   ja:RDFDataset .
text:TextIndexLucene  rdfs:subClassOf   text:TextIndex .

[] rdf:type fuseki:Server ;
   fuseki:services (
     <#service_text_tdb>
   ) .

<#service_text_tdb> rdf:type fuseki:Service ;
    rdfs:label                      "TDB/text service" ;
    fuseki:name                     "ds" ;
    fuseki:serviceQuery             "query" ;
    fuseki:serviceQuery             "sparql" ;
    fuseki:serviceUpdate            "update" ;
    fuseki:serviceUpload            "upload" ;
    fuseki:serviceReadGraphStore    "get" ;
    fuseki:serviceReadWriteGraphStore    "data" ;
    fuseki:dataset                  :text_dataset ;
.

:text_dataset rdf:type     text:TextDataset ;
    text:dataset   <#dataset> ;
    text:index     <#indexLucene> ;
    .

<#dataset> rdf:type      tdb:DatasetTDB ;
    tdb:location "/tmp/db" ;
    tdb:unionDefaultGraph true ; # Optional
    .

<#indexLucene> a text:TextIndexLucene ;
    text:directory <file:/tmp/lucene> ;
    text:entityMap <#entMap> ;
    text:storeValues true ;
    text:defineAnalyzers (
      [ 
        text:defineAnalyzer <#configuredAnalyzer> ;
        text:analyzer [
          a text:ConfigurableAnalyzer ;
          text:tokenizer <#tokenizer> ;
          text:filters ( :selectiveFoldingFilter text:LowerCaseFilter )
        ]
      ]
      [
        text:defineTokenizer <#tokenizer> ;
        text:tokenizer [
          a text:GenericTokenizer ;
          text:class "org.apache.lucene.analysis.core.LowerCaseTokenizer" 
        ]
      ]
      [
        text:defineFilter :selectiveFoldingFilter ;
        text:filter [
          a text:GenericFilter ;
          text:class "org.apache.jena.query.text.filter.SelectiveFoldingFilter" ;
          text:params (
            [ 
              text:paramName "whitelisted" ;
              text:paramType text:TypeSet ;
              text:paramValue ("ç" "ä")
            ]
          )
        ]
      ]
    ) ;
    text:analyzer [
      a text:DefinedAnalyzer ;
      text:useAnalyzer <#configuredAnalyzer> 
    ] ;
    text:queryAnalyzer [ 
      a text:DefinedAnalyzer ;
      text:useAnalyzer <#configuredAnalyzer> 
    ] ;
    text:queryParser text:AnalyzingQueryParser ;
    text:multilingualSupport true ;
 .

<#entMap> a text:EntityMap ;
    text:defaultField     "pref" ;
    text:entityField      "uri" ;
    text:uidField         "uid" ;
    text:langField        "lang" ;
    text:graphField       "graph" ;
    text:map (
         # skos:prefLabel
         [ text:field "pref" ;
           text:predicate skos:prefLabel
         ]
         # skos:altLabel
         [ text:field "alt" ;
           text:predicate skos:altLabel
         ]
         # skos:hiddenLabel
         [ text:field "hidden" ;
           text:predicate skos:hiddenLabel 
         ]
     ) 
 .
```

**Comment by kinow** [comment]
Then, started Fuseki in Eclipse (FusekiCmd, with --config /.../fuseki.ttl). Loading the data file on to the /ds/ endpoint, everything works as expected. I loaded a modified `books.ttl` and got the following:

**Comment by kinow** [comment]
screenshot_2018-04-13_00-51-46
screenshot_2018-04-13_00-52-12

**Comment by kinow** [comment]
As it is possible to see, that as the configuration white-lists only `ä`, the `ö` is escaped with the filter.

**Comment by kinow** [comment]
Used `luke` to look at the Lucene index created, and everything checked. Had a bit of struggle with the queries, but it was my mistake.

File: jena-text/src/test/java/org/apache/jena/query/text/filter/TestSelectiveFoldingFilter.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+     *
+     * @param whitelisted white-list
+     * @return list of CharTermAttribute converted to a list of String's
+     * @throws IOException from Lucene API
+     */
+    private List<String> collectTokens(StringReader inputText, CharArraySet whitelisted) throws IOException {
+        StandardTokenizer tokenizer = new StandardTokenizer();
+        tokenizer.setReader(inputText);
+
+        // Replace this by a C, and then the ASCIIFolding function will take care to replace it by c, as it was not whitelisted
```
Comment (rvesse) [inline jena-text/src/test/java/org/apache/jena/query/text/filter/TestSelectiveFoldingFilter.java]: Leftover dev comment?

Comment (kinow) [reply jena-text/src/test/java/org/apache/jena/query/text/filter/TestSelectiveFoldingFilter.java]: You got me @rvesse . Spent some 5 minutes looking at the code... at the comment... at the code... at the comment... and couldn't remember every writing it to be honest :-) definitely some leftover. Nice catch, thanks for reviewing it.

**Comment by osma** [comment]
I tested this locally and it seems to work as it should. Great work!

**Comment by kinow** [comment]
Thanks for testing it @osma ! If there are no objections until next weekend I will merge it. With 3.7.0 out, we probably have some more time for testing, and then update the documentation. In the meantime, will jump back to JENA-632 :D

**Comment by kinow** [comment]
Merged in a56f7710d88e368824042863e1ebef9afc7fd5f3, closing (forgot to use the magic words in the commit to close this PR automatically)

**Pull request closed without merging**